drivers/flash/soc_flash_numaker_rmc.c

```diff
@@ -0,0 +1,275 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (c) 2024 Nuvoton Technology Corporation.
+ */
+
+#define DT_DRV_COMPAT nuvoton_numaker_rmc
+
+#include <string.h>
+#include <errno.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/logging/log.h>
+#include "flash_priv.h"
+#include <NuMicro.h>
+
+LOG_MODULE_REGISTER(flash_numaker_rmc, CONFIG_FLASH_LOG_LEVEL);
+
+#define SOC_NV_FLASH_NODE DT_INST(0, soc_nv_flash)
+#define SOC_NV_FLASH_WRITE_BLOCK_SIZE DT_PROP_OR(SOC_NV_FLASH_NODE, write_block_size, 0x04)
+
+struct flash_numaker_data {
+	RMC_T *rmc;
+	struct k_sem write_lock;
+	uint32_t flash_block_base;
+};
+
+static const struct flash_parameters flash_numaker_parameters = {
+	.write_block_size = SOC_NV_FLASH_WRITE_BLOCK_SIZE,
+	.erase_value = 0xff,
+};
+
+/* Validate offset and length */
+static bool flash_numaker_is_range_valid(off_t offset, size_t len)
+{
+	uint32_t aprom_size = (RMC_APROM_END - RMC_APROM_BASE);
+
+	/* check for min value */
+	if ((offset < 0) || (len == 0)) {
+		return false;
+	}
+
+	/* check for max value */
+	if (offset >= aprom_size || len > aprom_size || (aprom_size - offset) < len) {
+		return false;
+	}
+
+	return true;
+}
+
+/*
+ * Erase a flash memory area.
+ *
+ * param dev       Device struct
+ * param offset    The address's offset
+ * param len       The size of the buffer
+ * return 0       on success
+ * return -EINVAL erroneous code
+ */
+
+static int flash_numaker_erase(const struct device *dev, off_t offset, size_t len)
+{
+	struct flash_numaker_data *dev_data = dev->data;
+	uint32_t rc = 0;
+	unsigned int key;
+	int page_nums = (len / RMC_FLASH_PAGE_SIZE);
+	uint32_t addr = dev_data->flash_block_base + offset;
+
+	/* return SUCCESS for len == 0 (required by tests/drivers/flash) */
+	if (!len) {
```

Comment:
nit: should probably be `if (len == 0) {` since len is not a boolean, same below